Project-Wide Refactor: Sessions, Firestore Architecture, and Role Utilities

MisbahAN opened the pull request:
This PR fully restructures how the app handles sessions, Firestore access, and role-specific logic. All Firestore path construction is centralized, the schema is cleaned up, and controllers/views now rely on clear utilities rather than raw Firestore calls.

MisbahAN committed:
Add global AppSession and remove obsolete SessionManager usage.

MisbahAN commented:
# Global UserSession Refactor

This PR introduces a new global `AppSession` singleton that replaces the old `SessionManager` system. The previous session flow relied on passing Firestore and user data through activities or re-querying Firestore repeatedly. This made the codebase harder to work with and unintuitive for developers.

## ✔️ What Changed

* Added `AppSession.java` — a global in-memory store for:
  * `currentUser`
  * `currentUserId`
  * `role`
  * shared Firestore instance (`AppSession.db()`)
* Removed `SessionManager.java` completely (no longer needed)
* Updated Splash/Login/Signup flows to initialize the session with:
```java
  AppSession.initialize(user);
```
* Updated all controllers to use:
```java
  this.firestore = AppSession.db();
```
  instead of creating their own instances.
* Removed unused `FirestoreManager.java`.

## ✔️ Why This Change Was Made

* Developers previously had no clear way to access the current user.
* User information was being fetched repeatedly in views and controllers.
* Firestore instances were being created in multiple places.
* Passing user info through Intents is error-prone and not scalable.

## ✔️ How Developers Should Use This Going Forward

To access current user:
```java
User user = AppSession.getUser();
```

To access role:
```java
String role = AppSession.getRole();
```

To access shared Firestore:
```java
FirebaseFirestore db = AppSession.db();
```

Developers do not need to store user data manually, pass it via intents, or re-query Firestore to know who is logged in.

MisbahAN committed:
Restructure Firestore schema for event images and waiting lists.

MisbahAN commented:
# Firestore Schema Restructure for Images & Waiting Lists

This step reorganizes all image and waiting-list data so it lives under its parent event, making the schema intuitive, consistent, and easy for developers to work with.

## ✔️ What Changed

### 1. Images moved under events/

New structure:
```
events/{eventId}/images/{poster|qrcode|...}
```

* All image-related Firestore operations (fetching, uploading, deleting) are now handled inside the role utilities:
  * `OrganizerDb` – upload/edit poster images
  * `AdminDb` – list/delete images across events
  * `EntrantDb` – fetch event images for entrant views
* `ImageController` no longer constructs paths or calls `AppSession.db()` directly.
* `Event` model gained helpers for retrieving poster/QR image IDs.

### 2. Waiting List moved under events/

New structure:
```
events/{eventId}/waitingList/{entrantId}
```

* `WaitingListController` no longer touches Firestore directly. It simply validates input and delegates the real work to:
  * `EntrantDb.joinWaitingList(...)`
  * `EntrantDb.leaveWaitingList(...)`
  * `OrganizerDb.fetchWaitingList(...)`
* Old root-level `waitingLists/` logic was completely removed.
* `UserController`'s entrant-history now uses:
```java
  collectionGroup("waitingList")
```
  to aggregate history across all events.

### 3. Views updated to use new nested schema

All UI components that show images or waiting lists now call the role utilities, not Firestore directly.

Updated screens include:
* `EventAdapter`
* `EventCardAdapter`
* `OrganizerToolsFragment`
* `AdminImageListFragment`
* `WaitingListActivity`
* `EventDetailsFragment`
* `OrganizerEventsFragment`
* `OrganizerEventsFragment` waiting-list listeners
* Entrant/Organizer dashboards

No fragment or controller constructs Firestore paths like `.collection("events")` manually anymore.

### 4. Tests and documentation updated

* `EventTest`, `ImageTest`, and `ImageControllerTest` updated to reflect nested image + waiting-list structure.
* `firebase-setup.md` updated so onboarding instructions match the new schema.

## ✔️ Why This Change Was Made

* Root-level `images/` and `waitingLists/` were confusing and did not match how events actually own their assets.
* By nesting under events, relationships became clear and consistent.
* Future features such as:
  * chosen entrants
  * cancelled entrants
  * history queries
  * event-scoped image cleanup become much easier and cleaner.
* Role utilities now encapsulate all path logic, so schema changes only need to be updated in one place.

## ✔️ Developer Usage Going Forward

Fetching images (entrant or organizer):
```java
EntrantDb.fetchEventImages(eventId);
OrganizerDb.fetchEventImage(eventId, imageId);
```

Uploading poster images:
```java
OrganizerDb.uploadPoster(eventId, imageModel);
```

Joining / leaving waiting list:
```java
EntrantDb.joinWaitingList(eventId, entrantId);
EntrantDb.leaveWaitingList(eventId, entrantId);
```

Fetching waiting list:
```java
OrganizerDb.fetchWaitingList(eventId);
```

## ❌ Forbidden going forward

Do not use root-level:
```
images/
waitingLists/
```

Do not manually construct Firestore paths in controllers/fragments.

**All Firestore access must go through the role utilities.**

MisbahAN committed:
Add role-specific Firestore utilities for Admin, Organizer, Entrant.

MisbahAN commented:
# Introduce Role-Based Firestore Utility Layer

This step adds a clean, intuitive API for all Firestore operations tied to specific user roles. Instead of controllers or fragments manually constructing Firestore paths, each role now gets its own high-level utility class.

This makes the codebase easier to understand, reduces repeated logic, and standardizes how Firestore is accessed across the app.

## ✔️ What Changed

New role utilities added under `utils/role/`:
* `AdminDb`
* `OrganizerDb`
* `EntrantDb`

Each class is a static utility that uses:
```java
AppSession.db()
```
for Firestore access and wraps all Firestore operations behind strongly typed helper methods.

## ✔️ What Each Utility Provides

### AdminDb
* List all events
* List all profiles
* Delete events and users
* Aggregate all images across events
* Delete event-scoped images safely

### OrganizerDb
* Create and update events
* Upload poster images to `events/{eventId}/images/{docId}`
* Fetch waiting lists
* Select random entrants for lottery
* Store chosen-entrant notifications under `events/{eventId}/chosen/{autoId}`

### EntrantDb
* Join or leave an event's waiting list
* Fetch event details
* Retrieve entrant history via `collectionGroup("waitingList")`

## ✔️ Why This Change Was Made

* Previously, every controller/view repeated Firestore path construction.
* Developers needed to remember nested Firestore paths manually.
* Responsibilities between roles weren't clearly separated.
* Adding new features required updating Firestore logic in many places.

Now:
* Firestore access is centralized and consistent.
* Developers get autocomplete for 90% of operations they need.
* Future changes to Firestore schemas won't break 20+ files — only role utilities.

## ✔️ How Developers Should Use This Going Forward

Organizer example:
```java
OrganizerDb.fetchWaitingList(eventId);
```

Entrant example:
```java
EntrantDb.joinWaitingList(eventId, entrantId);
```

Admin example:
```java
AdminDb.listAllImagesAcrossEvents();
```

**Developers should no longer construct Firestore paths manually in controllers or fragments.**

MisbahAN committed:
Route controllers and views through role-based database utilities.

MisbahAN commented:
# Wire Controllers and Views to Role Database Utilities

This step removes direct Firestore usage from controllers and UI classes and routes everything through the new `AdminDb`, `OrganizerDb`, and `EntrantDb` utilities.

## ✔️ What Changed

* **Controllers now delegate all Firestore operations:**
  * `EventController` uses:
    * `OrganizerDb` for event creation/updates and event listeners
    * `EntrantDb` / `AdminDb` for event reads where appropriate
  * `ImageController` calls organizer/admin/entrant image helpers instead of building `events/{eventId}/images/...` paths itself.
  * `UserController`:
    * Uses `EntrantDb` for profile CRUD and history
    * Uses `AdminDb` for admin-only profile operations
  * `WaitingListController` forwards join/leave/fetch logic to `EntrantDb`.

* **Role DBs were expanded to support these call sites:**
  * `EntrantDb` now exposes helpers for:
    * Listing/filtering events
    * Fetching event images
    * Profile CRUD and history
    * Login/signup flows and device binding
  * `OrganizerDb` now supports:
    * Fetching and listening to organizer events
    * Listening to waiting lists
    * Managing event images
    * Cleaning up device bindings on logout
  * `AdminDb` now:
    * Lists/deletes events and profiles
    * Aggregates and deletes images across events
    * Handles device unbinding for admin-driven removals

* **Activities/fragments no longer construct Firestore paths directly:**
  * `AdminActivity`, `OrganizerActivity`, and `EntrantActivity` use their role DBs for logout/device cleanup instead of custom Firestore logic.
  * `AdminImageListFragment` uses `AdminDb` to aggregate and delete event-scoped images.
  * `OrganizerEventsFragment` and `WaitingListActivity` subscribe via `OrganizerDb` instead of calling `AppSession.db()` directly.
  * `LoginActivity`, `SignupActivity`, and `SplashActivity` query and modify users exclusively through `EntrantDb` and then initialize `AppSession`.

## ✔️ Why This Change Was Made

* Previously, controllers and fragments were manually constructing Firestore paths and mixing business logic with database details.
* Any schema change required edits in many places.
* It wasn't obvious where to look for "how does X role talk to the database?"

Now:
* Firestore path construction is fully centralized in the role DB layer.
* Controllers and views read much more cleanly and focus on orchestration + UI, not low-level Firestore details.
* Future schema changes or optimizations can be handled inside `AdminDb` / `OrganizerDb` / `EntrantDb` with minimal impact on the rest of the app.

## ✔️ How Developers Should Use This Going Forward

* **Organizers:** use `OrganizerDb` for all event and waiting-list operations.
* **Entrants:** use `EntrantDb` for joining/leaving waiting lists, fetching events, and profile/history operations.
* **Admins:** use `AdminDb` for listing/deleting events, profiles, and images.

**No new code in controllers or views should call `AppSession.db().collection(...)` or construct Firestore paths manually.**

MisbahAN committed:
Use: Update EventControllerTest for new API.

MisbahAN committed:
Clean unused helpers and remove legacy waiting-list code.

MisbahAN commented:
# Cleanup of Legacy Firestore Logic and Unused Code

This step removes leftover helper methods and tests that became obsolete after routing all Firestore access through the role-based utilities.

## ✔️ What Changed

* Removed the now-unused helper methods from `WaitingListController`:
  * `mapSnapshot`
  * `buildWaitingListData`
  * other Firestore path scaffolding
* Removed the old `WaitingListControllerTest`, which only tested the deleted helpers and no longer matched the controller's actual behavior.
* Cleaned up unused imports created by previous steps.

## ✔️ Why This Change Was Made

After Steps A–D:
* All Firestore interactions moved into `EntrantDb`, `OrganizerDb`, and `AdminDb`.
* Controllers no longer maintain their own Firestore path-building logic.
* The deleted helpers were no longer reachable and only created confusion.

This cleanup ensures the codebase stays maintainable and avoids dead code that could mislead developers.

## ✔️ Invariant Check

Controllers and views no longer:
* Call `FirebaseFirestore.getInstance()`
* Call `FirebaseUtils.getFirestore()`
* Manually write `AppSession.db().collection(...)`

**All Firestore access is routed exclusively through the role-specific utilities.**

MisbahAN committed:
Add comprehensive Javadocs across session, role utilities, controllers, models, and UI.

MisbahAN committed:
Add Gradle task for generating Android Javadocs and docs.

MisbahAN commented:
## Generating Javadocs

The latest commit adds a Gradle task for generating Android-aware Javadocs.

**Generate documentation:**
```bash
./gradlew generateAndroidJavadocs
```

Output is written to `/javadocs/`. Run this command from the project root.

This change only adds documentation tooling—no build behavior, dependencies, or runtime logic were modified.

Orion-W commented on app/src/main/java/com/example/impact/utils/AppSession.java:
```diff
@@ -0,0 +1,80 @@
+     * @return shared Firestore instance
+     */
+    public static FirebaseFirestore db() {
+        return db;
```
This is not needed. The User Session should not store database related attributes

MisbahAN replied:
I've already explained the reasoning behind AppSession containing `db()` in my reply below - but to summarize briefly here:

I wasn't treating it as a "user session," but as a broader AppSession containing global state the app depends on (current user + Firestore entry point).

That said, if we prefer keeping DB access exclusively in `FirebaseUtils` instead of centralizing it under AppSession, I'll remove the `db()` getter and revert this part of the change. lmk what you think.

Orion-W commented on app/src/main/java/com/example/impact/utils/AppSession.java:
This is a good implementation of the Session system we talked about yesterday. I don't want the AppSession to have a db object associated with it, it seems redundant as the exact same thing is happening in the FirebaseUtils class.

MisbahAN replied:
# About AppSession containing `db()`

I understand the concern.

Just to explain my thought process: I didn't see this as a "UserSession" anymore, I saw it as a central AppSession that stores everything belonging to the **running instance of the app** (session state). The database and the current user **both** fall under that umbrella:

* both are globally-required state
* both are conceptually part of the "app session lifecycle"
* both are accessed across 10 - 20 classes regularly

So my reasoning was:

If we already have a single in-memory AppSession that exposes user, it feels natural for it to expose the app's DB entry point as well.

If you preferred this direction, I was eventually planning to remove `FirebaseUtils` entirely so we'd just have:
```java
AppSession.db()
AppSession.getUser()
AppSession.getRole() (optional)
```

and everything lives under one unified "app state" API instead of being split across multiple globals.

## Examples of how AppSession was meant to be used

**Get current user email**
```java
String email = AppSession.getUser().getEmail();
```

**Get Firestore**
```java
FirebaseFirestore db = AppSession.db();
```

This design keeps the app's "active session state" together.

---

That said, if we want DB access to exclusively remain in `FirebaseUtils`, I can remove the db() getter and revert that part. LMK the direction you prefer.

jeremycarefoot commented:
- Why do we need an AdminDB, EntrantDB, and OrganizerDB? These classes seem redundant, and I don't see any problem with our previous pattern of the controllers. In my opinion, alot of this is over-abstraction. We have numerous coupled classes that interact with the DB which just overconfuses things for me.
- Why do we have empty private constructors for several classes? Unless I'm missing something this is just dead code

The rest LGTM (image storage/AppSession stuff).

Orion-W commented on app/src/main/java/com/example/impact/view/AdminActivity.java:
```diff
@@ -35,9 +38,10 @@ public class AdminActivity extends AppCompatActivity {
-
-        adminId = getIntent().getStringExtra(LoginActivity.EXTRA_USER_ID);
+        adminId = AppSession.getUserId();
+        if (adminId == null) {
```
This check logic should be in the AppSession method not here in the activity.

MisbahAN replied:
just pushed an update for this.

I moved all session-validity checks (like verifying `userId` isn't null) into `AppSession` itself, so Activities/Fragments now simply call:
```
AppSession.getUserId();
AppSession.getUser();
```

without doing any extra safety checks on their side.

MisbahAN commented:
@jeremycarefoot 

> * Why do we need an AdminDB, EntrantDB, and OrganizerDB? These classes seem redundant, and I don't see any problem with our previous pattern of the controllers. In my opinion, alot of this is over-abstraction. We have numerous coupled classes that interact with the DB which just overconfuses things for me.

The role-specific DB classes fix a few concrete issues we had before:

**Previously:** Controllers were manually building Firestore paths everywhere, duplicating logic across files, and hard-coding nested paths like `events/{id}/waitingList/...`. Schema changes required updating 20–30 places.

**Now:**
* Firestore paths are centralized in 3 files instead of scattered everywhere
* Controllers just call `OrganizerDb.createEvent(...)` or `EntrantDb.joinWaitingList(...)` without knowing the schema
* Each role's distinct behavior is cleanly separated

---

> * Why do we have empty private constructors for several classes? Unless I'm missing something this is just dead code

I looked through the difference again and didn’t see any empty classes. If you can point out which ones you’re referring to, I’ll remove or clean them up right now.

Orion-W commented on app/src/main/java/com/example/impact/view/AdminImageListFragment.java:
```diff
@@ -85,8 +86,14 @@ public void onDeleteImageClicked(int position, Image image) {
                 () -> {
-                    imageController.deleteImage(imageId, v -> onImageDelete(imageFilename),
-                            error -> Toast.makeText(getContext(), R.string.admin_image_list_error_deletion, Toast.LENGTH_SHORT).show());
+                    String eventId = image.getEventId();
```
This Image delete listener should definitely call the ImageController to manage the deletion of the image from the database.

MisbahAN replied:
Fixed - the UI no longer calls `AdminDb` directly. Deletion is now routed entirely through `ImageController.deleteImage(...)`, which internally calls the correct RoleDb method. This keeps the controller as the single entry point for all image operations.

Orion-W commented on app/src/main/java/com/example/impact/view/AdminImageListFragment.java:
```diff
@@ -85,8 +86,14 @@ public void onDeleteImageClicked(int position, Image image) {
+                        return;
+                    }
+                    AdminDb.deleteEventImage(eventId, imageId)
+                            .addOnSuccessListener(unused -> onImageDelete(imageFilename))
```
In future the success listener should be passed to the ImageController.

MisbahAN replied:
Updated - success/error listeners are now passed into `ImageController`, not the UI. The fragment no longer touches `AdminDb.deleteEventImage(...)`; everything flows through `ImageController`, which handles the RoleDb calls and listener wiring.

Orion-W commented on app/src/main/java/com/example/impact/view/EntrantActivity.java:
```diff
@@ -2,40 +2,46 @@
+        User user = AppSession.getUser();
+        entrantId = user != null ? user.getId() : null;
+        if (TextUtils.isEmpty(entrantId)) {
+            entrantId = getIntent().getStringExtra(LoginActivity.EXTRA_USER_ID);
```
Again lets have this kind of check happen in the AppSession class not in views

MisbahAN replied:
moved all of that logic into AppSession, so the view no longer performs any fallback checks.

MisbahAN committed:
Move all userId validation into AppSession; simplify Admin/Entrant Activities.

Orion-W commented on app/src/main/java/com/example/impact/view/EntrantActivity.java:
```diff
@@ -114,12 +120,29 @@ public boolean onOptionsItemSelected(MenuItem item) {
-            startActivity(intent);
-            finish();
-        });
+        String deviceId = Settings.Secure.getString(getContentResolver(), Settings.Secure.ANDROID_ID);
```
This device session code should be added to the AppSession class. Such that we can just do:
```
try {
    AppSession.logout();
    // Success toast
} catch (Exception e) {
    // Error toast
}
navigateToLogin();
```
Since the success toast doesn't seem to need to be defined in the success listener of clearDeviceBinding() we should move it out of the function call and instead check the success of the clearDeviceBinding using some attribute or method in AppSession. Or maybe using error handling like above.

jeremycarefoot commented on app/src/main/java/com/example/impact/utils/AppSession.java:
```diff
@@ -0,0 +1,114 @@
+    private static String currentUserId;
+    private static String role;
+
+    private AppSession() {
```
Dead constructor?

jeremycarefoot commented on app/src/main/java/com/example/impact/utils/role/AdminDb.java:
```diff
@@ -0,0 +1,171 @@
+    private static final FirebaseFirestore db = AppSession.db();
+
+    private AdminDb() {
+        // no-op
```
Dead constructor?

jeremycarefoot commented on app/src/main/java/com/example/impact/utils/role/OrganizerDb.java:
```diff
@@ -0,0 +1,293 @@
+    private static final FirebaseFirestore db = AppSession.db();
+
+    private OrganizerDb() {
+        // no instances
```
Dead constructor?

Orion-W commented on app/src/main/java/com/example/impact/view/LoginActivity.java:
```diff
@@ -92,29 +87,35 @@ private void attemptLogin() {
+     * @param user mapped user model
      */
-    private void navigateToRole(String role, String userId, @Nullable String email) {
+    private void navigateToRole(@Nullable String role, @NonNull User user) {
```
Good method to have

Orion-W commented on app/src/main/java/com/example/impact/view/LoginActivity.java:
```diff
@@ -127,9 +128,9 @@ private void navigateToRole(String role, String userId, @Nullable String email)
-        intent.putExtra(EXTRA_USER_ID, userId);
-        if (email != null) {
-            intent.putExtra("extra_user_email", email);
+        intent.putExtra(EXTRA_USER_ID, user.getId());
```
Lets handle intent logic like this in AppSession

MisbahAN committed:
Route all image fetch/upload/delete flows through ImageController.

Orion-W reviewed (changes requested):
Unfortunately I can't sign off on this PR.

Orion-W commented:
I made a branch last night refactoring some of the model controllers to use the FirebaseUtils class. 
While I was doing that I realized that doing that is mostly redundant because the functionality of the dev facing methods stays the same. 
A FirebaseUtils class would have been much more useful in the first week of this project but its implementation now is almost not worth the time.

In regards to this PR I feel that what I said about FirebaseUtils applies to the classes in role directory (AdminDB, ...).
I think that the structure of these classes is not intuitive.
Going forward I would like the data model controllers to directly call FirebaseUtils for interaction with the database.

MisbahAN commented:
thanks for the feedback @Orion-W 

Yesterday we discussed the direction of centralizing Firestore paths and session logic, so I moved forward with that architecture and refactored the codebase accordingly. The AppSession + RoleDb abstraction was built based on that discussion - the goal was to reduce schema duplication, simplify controllers, and make Firestore interactions more consistent.

That said, I understand your point about preferring a flatter structure where controllers call FirebaseUtils directly.

I’ve been working on this for hours, so I’m going to take a break and get back to it later tomorrow. If you have any tasks you want me to handle, lmk - or if you think the team should meet to decide the direction, feel free to ping everyone in Discord :)
